airflow/decorators/base.py

```diff
@@ -176,11 +178,110 @@ def _hook_apply_defaults(self, *args, **kwargs):
 
 T = TypeVar("T", bound=Callable)
 
+OperatorSubclass = TypeVar("OperatorSubclass", bound="BaseOperator")
+
+
+@attr.define
+class OperatorWrapper(Generic[T, OperatorSubclass]):
+    """
+    Helper class for providing dynamic task mapping to decorated functions.
+
+    ``task_decorator_factory`` returns an instance of this, instead of just a plain wrapped function.
+
+    :meta private:
+    """
+
+    function: T = attr.ib(validator=attr.validators.is_callable())
+    operator_class: Type[OperatorSubclass]
+    multiple_outputs: bool = attr.ib()
+    kwargs: Dict[str, Any] = attr.ib(factory=dict)
+
+    decorator_name: str = attr.ib(repr=False, default="task")
+    function_arg_names: Set[str] = attr.ib(repr=False)
+
+    @function_arg_names.default
+    def _get_arg_names(self):
+        return set(inspect.signature(self.function).parameters)
```

Comment:
Does this need to exist? From what I can tell it is only used once in the validator (which already has access to the original `function`).

Reply:
`@function_arg_names.default` decorator means that this is the default value for attrs.

No this doesn't need to exist as a function, it could possible be a lamda in `function_arg_names = attr.ib()` call.

Reply:
I think a function is clearer than this:

```
    function_arg_names: Set[str] = attr.ib(
        repr=False,
        default=attr.Factory(lambda i: set(inspect.signature(i.function).parameters), takes_self=True),
    )
```

Reply:
One option is I could do `del _get_arg_names` -- as once it's decorated the function doesn't need to remain on the class.

Reply:
This would be the change that addresses most of your points. What do you think?

```diff
diff --git a/airflow/decorators/base.py b/airflow/decorators/base.py
index 3ee13aed4..6f1fa53b9 100644
--- a/airflow/decorators/base.py
+++ b/airflow/decorators/base.py
@@ -28,6 +28,7 @@ from airflow.models.baseoperator import BaseOperator, MappedOperator
 from airflow.models.dag import DAG, DagContext
 from airflow.models.xcom_arg import XComArg
 from airflow.utils.task_group import TaskGroup, TaskGroupContext
+from airflow.compat.functools import cached_property
 
 
 def validate_python_callable(python_callable):
@@ -181,7 +182,7 @@ T = TypeVar("T", bound=Callable)
 OperatorSubclass = TypeVar("OperatorSubclass", bound="BaseOperator")
 
 
-@attr.define
+@attr.define(slots=False)
 class OperatorWrapper(Generic[T, OperatorSubclass]):
     """
     Helper class for providing dynamic task mapping to decorated functions.
@@ -199,22 +200,32 @@ class OperatorWrapper(Generic[T, OperatorSubclass]):
     decorator_name: str = attr.ib(repr=False, default="task")
     function_arg_names: Set[str] = attr.ib(repr=False)
 
+    @cached_property
+    def function_signature(self):
+        return inspect.signature(self.function)
+
     @function_arg_names.default
     def _get_arg_names(self):
-        return set(inspect.signature(self.function).parameters)
+        return set(self.function_signature.parameters)
+
+    del _get_arg_names
 
     @function.validator
     def _validate_function(self, _, f):
         if 'self' in self.function_arg_names:
             raise TypeError(f'@{self.decorator_name} does not support methods')
 
+    del _validate_function
+
     @multiple_outputs.default
     def _infer_multiple_outputs(self):
-        sig = inspect.signature(self.function).return_annotation
+        sig = self.function_signature.return_annotation
         ttype = getattr(sig, "__origin__", None)
 
         return sig is not inspect.Signature.empty and ttype in (dict, Dict)
 
+    del _infer_multiple_outputs
+
     def __attrs_post_init__(self):
         self.kwargs.setdefault('task_id', self.function.__name__)
 
```

Reply:
Oh, I meant the `function_arg_names` attribute might not need to exist entirely, since it is only accessed by `_validate_function` (which is a validator for the `function` attribute) and the validator is most likely only run once anyway.

Reply:
Ohhh! I thought I used it in .map to validate args then

Reply:
It's used in `_validate_arg_names` (which is called from `.map` and `.partial`, and in the function validator, so it probably should exist in some form. It could be a cached property rather than an attr + default.

Reply:
See 39b42fa49